### PR TITLE
OKE-27032 Flannel Support

### PR DESCRIPTION
**Pull request by Inbaraj-S**
Added support for flannel CNI. Accepts Nodeport Service for Flannel CNI and adds nodes as backend based on externalTrafficPolicy set on service.
Added a new parameter in deployment - `cluster-id` to check for cluster details.

### CODE REVIEW

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+	"github.com/oracle/oci-native-ingress-controller/pkg/util"
+)
+
+var errIngressClassNotReady = errors.New("ingress class not ready")
+
+const (
+
+	// ServiceAnnotationLoadBalancerNodeFilter is a service annotation to select specific nodes as your backend in the LB
+	// based on label selector.
+	ServiceAnnotationLoadBalancerNodeFilter = "oci.oraclecloud.com/node-label-selector"
```
Comment (nirpai) [inline pkg/controllers/nodeBackend/nodeBackend.go:38]: Remove this.

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+		podLister:          podLister,
+		nodeLister:         nodeLister,
+		client:             client,
+		queue:              workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(10*time.Second, 5*time.Minute)),
+	}
+
+	return c
+}
+
+func (c *Controller) processNextItem() bool {
```
Comment (nirpai) [inline pkg/controllers/nodeBackend/nodeBackend.go:78]: Refactor this function (any other similar functions). This is repeated in all controllers.

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+				svcName, svcPort, _, svc, err := c.pathToServiceAndTargetPort(ingress.Namespace, path)
+				if err != nil {
+					return err
+				}
+
+				if svc == nil || svc.Spec.Ports == nil || svc.Spec.Ports[0].NodePort == 0 {
+					continue
+				}
+
+				pods, err := c.retrievePods(err, ingress, svcName, svcPort)
```
Comment (nirpai) [inline pkg/controllers/nodeBackend/nodeBackend.go:193]: Refactor retrieve pod in local case.

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+				}
+				backendSetName := util.GenerateBackendSetName(ingress.Namespace, svcName, svcPort)
+				err = c.client.GetLbClient().UpdateBackends(context.TODO(), lbID, backendSetName, backends)
+				if err != nil {
+					return fmt.Errorf("unable to update backends for %s/%s: %w", ingressClass.Name, backendSetName, err)
+				}
+			}
+		}
+	}
+	// Sync default backends
```
Comment (nirpai) [inline pkg/controllers/nodeBackend/nodeBackend.go:226]: Add nodes those host default backend pods.

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 5]
```diff
@@ -0,0 +1,496 @@
...
+	c.syncDefaultBackend(lbID, ingresses)
+	return nil
+}
+
+func (c *Controller) retrievePods(err error, ingress *networkingv1.Ingress, svcName string, svcPort int32) ([]*corev1.Pod, error) {
```
Comment (nirpai) [inline pkg/controllers/nodeBackend/nodeBackend.go:231]: Refactor pod retrieval to a generic place.

File: pkg/server/server.go [hunk 6]
```diff
@@ -113,7 +104,31 @@ func SetUpControllers(opts types.IngressOpts, ingressClassInformer networkinginf
 		go ingressClassController.Run(3, ctx.Done())
 		go ingressController.Run(3, ctx.Done())
 		go routingPolicyController.Run(3, ctx.Done())
-		go backendController.Run(3, ctx.Done())
+
+		if opts.CniType == "FLANNEL_OVERLAY" {
```
Comment (nirpai) [inline pkg/server/server.go:108]: Can we use ternary operator?

File: pkg/controllers/nodeBackend/nodeBackend.go (hunk 1 above)
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go:38]: Done.

File: pkg/controllers/nodeBackend/nodeBackend.go (hunk 2 above)
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go:78]: Done for both the backend controllers.

File: pkg/controllers/nodeBackend/nodeBackend.go (hunk 3 above)
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go:193]: Done

File: pkg/controllers/nodeBackend/nodeBackend.go (hunk 4 above)
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go:226]: Done

File: pkg/controllers/nodeBackend/nodeBackend.go (hunk 5 above)
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go:231]: Done

File: pkg/server/server.go (hunk 6 above)
Comment (Inbaraj-S) [reply pkg/server/server.go:108]: Done

File: pkg/controllers/nodeBackend/nodeBackend.go [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+	return nil
+}
+
+func (c *Controller) getNodeConditionPredicate() NodeConditionPredicate {
+	return func(node *corev1.Node) bool {
+
+		// Remove nodes that are about to be deleted by the cluster autoscaler.
+		for _, taint := range node.Spec.Taints {
+			if taint.Key == ToBeDeletedTaint {
+				klog.V(4).Infof("Ignoring node %v with autoscaler taint %+v", node.Name, taint)
```
[the remark this answers is not in the document]
Comment (Inbaraj-S) [reply pkg/controllers/nodeBackend/nodeBackend.go]: Done

**Review by nirpai — APPROVED** [review]

**Review by nirpai — APPROVED** [review]

**Pull request merged**